src/passes/MergeFunctions.cpp

```diff
@@ -0,0 +1,731 @@
+// operand of const and call insts) by parameterization.
+// Performing this pass at post-link time can merge more functions across
+// objects. Inspired by Swift compiler's optimization which is derived from
+// LLVM's one:
+// https://github.com/apple/swift/blob/main/lib/LLVMPasses/LLVMMergeFunctions.cpp
+// https://github.com/llvm/llvm-project/blob/main/llvm/docs/MergeFunctions.rst
+//
+// e.g.
+//
+//  Before:
+//    (func $big-const-42 (result i32)
+//      [[many instr 1]]
+//      (i32.const 44)
+//      [[many instr 2]]
+//    )
+//    (func $big-const-43 (result i32)
+//      [[many instr 1]]
+//      (i32.const 45)
+//      [[many instr 2]]
+//    )
+//  After:
+//    (func $byn$mgfn-shared$big-const-42 (result i32)
+//      [[many instr 1]]
+//      (local.get $0)
+//      [[many instr 2]]
+//    )
+//    (func $big-const-42 (result i32)
+//      (call $byn$mgfn-shared$big-const-42
+//       (i32.const 42)
+//      )
+//    )
+//    (func $big-const-43 (result i32)
+//      (call $byn$mgfn-shared$big-const-42
+//       (i32.const 43)
+//      )
+//    )
+
+#include "ir/manipulation.h"
+#include "ir/module-utils.h"
+#include "ir/names.h"
+#include "ir/utils.h"
+#include "opt-utils.h"
+#include "pass.h"
+#include "support/hash.h"
+#include "support/utilities.h"
+#include "wasm.h"
+#include <algorithm>
+#include <cassert>
+#include <cstddef>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <ostream>
+#include <variant>
+#include <vector>
+
+#if 0
+#define MERGEFUNC_DEBUG(X)                                                     \
+  do {                                                                         \
+    X;                                                                         \
+  } while (0)
+#else
+#define MERGEFUNC_DEBUG(X)                                                     \
+  do {                                                                         \
+  } while (0)
+#endif
+
+namespace wasm {
+
+// A function hash is calculated by considering only the order of instruction
+// types, the number of locals, and signatures Used by early classification of
+// functions to be merged.
+struct CheapHasher : public WalkerPass<PostWalker<CheapHasher>> {
+
+  bool isFunctionParallel() override { return true; }
+
+  using Map = std::map<Function*, size_t>;
+  Map* output;
+
+  CheapHasher(Map* output) : output(output) {}
+
+  CheapHasher* create() override { return new CheapHasher(output); }
+
+  static Map createMap(Module* module) {
+    Map hashes;
+    for (auto& func : module->functions) {
+      // ensure an entry for each function - we must not modify the map shape in
+      // parallel, just the values
+      hashes[func.get()] = hash(0);
+    }
+    return hashes;
+  }
+
+  void doWalkFunction(Function* func) { output->at(func) = hashFunction(func); }
+
+  static size_t hashFunction(Function* func) {
+    size_t digest = 0;
+    for (Type param : func->getParams()) {
+      rehash(digest, param);
+    }
```

Comment:
It is not correct to hash params like this - instead, the type should be hashed. In nominal typing in wasm GC the params may match but the type may cause them to differ.

That is handled correctly in this existing code:

https://github.com/WebAssembly/binaryen/blob/3168fa227eeb3b964b6d3c773d95e0a2014b396a/src/ir/hashed.h#L51-L53

It might be good to refactor so that code can be shared and not duplicated.